bpf/bpf_host.c

```diff
@@ -916,6 +916,10 @@ handle_netdev(struct __ctx_buff *ctx, const bool from_host)
 {
 	__u16 proto;
 
+	/* Pass vlan tagged packets to the stack - kernel will strip vlan tag and we'll receive it back on proper interface */
+	if (!from_host && ctx->vlan_present)
+		return CTX_ACT_OK;
```

Comment:
Could you describe some more the expected packet path with this commit for vlan tagged packets? Maybe via ascii drawing. The comment is a little too terse to judge.

Reply:
wan0 - original interface wan0.4000 - vlan interface, attached to wan0.

BPF programs attached on both wan0 and wan0.4000 interfaces.
wan0.4000 interface is used as primary (internal) network.
PT - tagged packet, P - normal untagged packet.

Original packet path:
```
* PT -> ingress on wan0
* PT -> cilium BPF ingress
* PT -> assigning pod's destination ip
* PT -> egress on wan0.4000
* PTT -> egress on wan0
```

Cilium makes no difference between tagged and untagged packets, cause the only difference with usual packets are 'vlan_*' fields which are not checked currently. As a result, packet received on wan0 will be load balanced to pod.
Packet will be sent through proper interface - wan0.4000 and kernel will add one more vlan tag for this packet and will pass it to egress on wan0.

Proper path (from my point of view):
```
* PT -> ingress on wan0
* PT -> pass back to kernel
* P -> ingress on wan0.4000
* P -> cilium BPF ingress
* P -> assigning pod's destination ip
* P -> egress on wan0.4000
* PT -> egress on wan0
```

I'm not very good at BPF, but I don't completely understand how it can affect BPF host routing or tunneling mode, cause the only place where vlan tagged packets can appear is physical interface currently. The idea here is: if we receive tagged packet then pass it back to kernel and kernel will do it's job - will strip vlan tag and will start processing packet on proper interface. And if we have vlan tagged packet on egress usually means that we've already processed this packet when it was untagged...

Reply:
Is it possible to configure the kernel in a way that after `* PT -> pass back to kernel` the packet bypasses bpf_host @ wan0.4000 ingress? If yes, then there is a security concern that the host firewall (implemented in bpf_host) can be bypassed.

Reply:
I think that there is no way for packet to bypass bpf_host@wan0.4000 ingress.

Reply:
> The idea here is: if we receive tagged packet then pass it back to kernel and kernel will do it's job - will strip vlan tag and will start processing packet on proper interface.

But then what happens if the NIC supports offloaded vlan, populates ctx->vlan_* fields (and has the tag already removed from the packet), then pushes up the stack with this PR, but there is no wan0.4000 device (or any other vlan typed netdev, fwiw)?

Reply:
The thing I'm a bit confused about is that almost all modern NICs will populate ctx->vlan_* fields via __vlan_hwaccel_put_tag() and read the vlan info from it's RX descriptor 
 (https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/net/ethernet/intel/ice/ice_txrx_lib.c#n179), so it's not part of the packet data to begin with. What NIC is backing your wan0?

Reply:
On real hardware I have 'Intel Corporation I210 Gigabit'. And I can see same behaviour on my VMs. Also there is no 802.1q header already in skb in handle_netdev function - only vlan_* fields. But I definitely should check what will happen in case there is no any/proper vlan typed device. Packet should be dropped, but I'm not sure.

Reply:
@borkmann , I have an answers for one question:

vlan packet is dropped by kernel always in case there is no proper vlan typed netdev. That means there is no security problem in case we will pass vlan targeted packets back to the kernel stack when we are processing them on main (physical) netdev. Such packets will be processed on proper vlan typed interface after that or will be dropped.

For now I'm investigating how it works with xdp and will get back. Thanks.